parallel.go

```diff
@@ -0,0 +1,143 @@
+package pail
+
+import (
+	"context"
+	"sync"
+
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+)
+
+type parallelBucketImpl struct {
+	Bucket
+	size         int
+	deleteOnSync bool
+	dryRun       bool
+}
+
+type ParallelBucketOptions struct {
+	Workers      int
+	DryRun       bool
+	DeleteOnSync bool
+}
+
+func NewParallelSyncBucket(opts ParallelBucketOptions, b Bucket) Bucket {
```

Comment:
then we have to add parallelism to a bunch of types, and the implementation is a bit annoying/hard to  get right.

Reply:
You can reuse these same functions, and keep them here in this file. All it would require would be if have an if statement checking the parallel sync flag, which if true would then use these functions. Just for sake of a consistent usage pattern.

Reply:
Or just check the option when constructing the bucket, and return the underlying bucket wrapped in `parallelSyncBucket`.

Reply:
The challenge is that the s3 bucket is a (popular) exception because you can push a lot (e.g. compression, size/etc.) to the api we're using, but really things like parallelism, and compression should probably be implemented as layers/wrappers.

Reply:
ah ok, so you would rather move to this type of usage pattern instead of adding options. sounds good!